File: genkit_backend/lib/flows/goal_coach_flow.dart
// ─────────────────────────────────────────────────────────────────────────────
// genkit_backend/lib/flows/goal_coach_flow.dart
//
// Genkit flow: goalCoach
// Multi-turn AI coaching conversation using Gemma 4 (cloud).
// ─────────────────────────────────────────────────────────────────────────────

import 'package:genkit/genkit.dart';
import 'package:genkit_google_genai/genkit_google_genai.dart';

GenkitFlow registerGoalCoachFlow(Genkit ai) {
  return ai.defineFlow(
    name: 'goalCoach',
    inputSchema: z.object({
      'userMessage':    z.string(),
      'goalTitle':      z.string(),
      'progressPercent': z.number(),
      'history': z.array(z.object({
        'role':    z.string(),
        'content': z.string(),
      })).optional(),
    }),
    outputSchema: z.object({
      'reply':             z.string(),
      'suggestedActions':  z.array(z.string()),
      'motivationalScore': z.number().int(),
    }),
    fn: (input) async {
      final userMessage     = input['userMessage'] as String;
      final goalTitle       = input['goalTitle'] as String;
      final progress        = (input['progressPercent'] as num).toDouble();
      final history         = (input['history'] as List<dynamic>? ?? []);

      // Build message history for multi-turn context
      final messages = <Message>[
        Message.system('''
You are the Goal Digger AI coach — supportive, concise, and action-oriented.
You help users break big goals into achievable tasks and stay motivated.
Current goal: "$goalTitle" (${(progress * 100).round()}% complete).
Reply in the same language as the user message.
Keep responses short (2–4 sentences) unless asked for detail.
After your reply, provide:
  - suggestedActions: 2–3 short action bullets the user could take right now
  - motivationalScore: integer 1–10 reflecting how motivated the user seems
Respond ONLY with valid JSON: { "reply": "...", "suggestedActions": [...], "motivationalScore": N }
'''),
        // Inject conversation history
        for (final msg in history)
          if ((msg as Map)['role'] == 'user')
            Message.user((msg)['content'] as String)
          else
            Message.model((msg)['content'] as String),
        // Current user message
        Message.user(userMessage),
      ];

      final response = await ai.generate(
        model: googleGenAI('gemma-4-it'),
        messages: messages,
        config: GenerationConfig(
          temperature: 0.7,
          maxOutputTokens: 512,
          responseMimeType: 'application/json',
        ),
      );

      final parsed = response.outputAsMap();
      return {
        'reply':             parsed['reply'] ?? '',
        'suggestedActions':  parsed['suggestedActions'] ?? [],
        'motivationalScore': parsed['motivationalScore'] ?? 7,
      };
    },
  );
}

File: functions/src/flows/goalCoachFlow.ts
// functions/src/flows/goalCoachFlow.ts
//
// Defines the goal coach conversational AI flow.
// The Flutter app calls this via the `goalCoach` Cloud Function.

import { z } from "genkit";
import { getAI, defaultModel } from "../ai";

const ChatMessageSchema = z.object({
  role: z.enum(["user", "model"]),
  content: z.string(),
});

const GoalCoachInputSchema = z.object({
  userMessage:          z.string(),
  goalTitle:            z.string(),
  progressPercent:      z.number().min(0).max(100).default(0),
  conversationHistory:  z.array(ChatMessageSchema).default([]),
});

const GoalCoachOutputSchema = z.object({
  reply:             z.string(),
  suggestedActions:  z.array(z.string()).default([]),
  encouragement:     z.string().default(""),
  // motivationalScore added so the Flutter model (which reads this field) works correctly.
  motivationalScore: z.number().min(1).max(10).default(7),
});

export type GoalCoachInput  = z.infer<typeof GoalCoachInputSchema>;
export type GoalCoachOutput = z.infer<typeof GoalCoachOutputSchema>;

export function defineGoalCoachFlow() {
  const ai = getAI();

  return ai.defineFlow(
    {
      name:         "goalCoach",
      inputSchema:  GoalCoachInputSchema,
      outputSchema: GoalCoachOutputSchema,
    },
    async (input) => {
      const historyBlock = input.conversationHistory
        .map((m) => `${m.role === "user" ? "User" : "Coach"}: ${m.content}`)
        .join("\n");

      const prompt = `
You are the Goal Digger AI coach — warm, concise, and relentlessly action-oriented.
The user's goal is: "${input.goalTitle}" (${Math.round(input.progressPercent)}% complete).

${historyBlock ? `Conversation so far:\n${historyBlock}\n` : ""}
User says: ${input.userMessage}

Respond with a JSON object:
{
  "reply": "1–3 sentence response. Be specific to the goal. Never generic.",
  "suggestedActions": ["concrete next task 1", "concrete next task 2"],
  "encouragement": "One short motivational line (max 12 words) tied to their goal",
  "motivationalScore": 8
}

Rules:
- suggestedActions should be specific micro-tasks the user can act on today.
- If the user asks to change/adjust tasks, put the revised task titles in suggestedActions.
- Reply must feel like a real human coach, not a chatbot.
- motivationalScore is 1–10: how encouraging the tone is (10 = very uplifting).
- Respond ONLY with valid JSON — no markdown fences.`.trim();

      try {
        const { text } = await ai.generate({
          model: defaultModel,
          prompt,
          config: {
            temperature: 0.65,
            maxOutputTokens: 512,
            responseMimeType: "application/json",
          },
        });

        // Strip potential markdown fences from response
        const cleaned = text
          .trim()
          .replace(/^```(?:json)?\s*/i, "")
          .replace(/\s*```\s*$/i, "")
          .trim();

        const parsed = JSON.parse(cleaned);

        return {
          reply:             String(parsed.reply ?? "").trim(),
          suggestedActions:  Array.isArray(parsed.suggestedActions)
            ? parsed.suggestedActions.map(String).filter(Boolean)
            : [],
          encouragement:     String(parsed.encouragement ?? "").trim(),
          motivationalScore: typeof parsed.motivationalScore === "number"
            ? Math.min(10, Math.max(1, Math.round(parsed.motivationalScore)))
            : 7,
        };
      } catch (e) {
        // Graceful degradation — the Flutter app handles this gracefully
        return {
          reply:             "I couldn't reach the AI right now. Check your network and try again.",
          suggestedActions:  [],
          encouragement:     "",
          motivationalScore: 7,
        };
      }
    }
  );
}
